File: 2156-find-substring-with-given-hash-value/2156-find-substring-with-given-hash-value.java
class Solution {
    public String subStrHash(String s, int power, int modulo, int k, int hashValue) {
        int n = s.length();
        long weight = 1;
        for(int i = 0; i < k - 1; i++) weight = (weight * power) % modulo;
        String res = "";
        int r = n - 1;
        long hash = 0;
        
        for(int i = n -1; i>=0; i--) {
            int val = s.charAt(i) - 'a' + 1;
            hash = (hash * power % modulo + val) % modulo;
            if(r - i + 1 == k) {
                if(hash == hashValue)
                    res = s.substring(i, r + 1);
                //remove higher order element
                hash = (hash + modulo - (s.charAt(r--) - 'a' + 1) * weight % modulo) % modulo;
            }
        }
        return res;
    }
}

File: 2156-find-substring-with-given-hash-value/2156-find-substring-with-given-hash-value.py
class Solution:
    def subStrHash(self, s: str, power: int, modulo: int, k: int, hashValue: int) -> str:
        n = len(s)
        weight = 1
        for i in range(k -1):
            weight = (weight * power) % modulo
        hash = 0
        r = n - 1
        res = ''
        for i in range(n - 1, -1, -1):
            curr = ord(s[i]) - ord('a') + 1
            hash = (hash*power%modulo + curr) % modulo
            if r - i + 1 == k:
                if hash == hashValue:
                    res = s[i:r+1]
                hash = (hash + modulo - (ord(s[r]) - ord('a') + 1)*weight%modulo) % modulo
                r -= 1
        return res
            